Add DAG calendar view endpoint

SavingFrame opened the pull request:
Hello everyone!

This PR adds a new DAG calendar view endpoint that provides scheduling information in a calendar format. This endpoint returns the calendar that existed in Airflow 2.x but provides additional functionality. The new endpoint includes filtering by `logical_date` and the ability to choose granularity, allowing users to customize their calendar view based on specific time periods and detail levels.

## Changes in this PR:
- Create service file that processes historical DagRuns and presumes queued runs
- Add UI endpoint
- Update Swagger documentation
- Add tests for different options of this endpoint

## How was this patch tested?

- Added comprehensive unit tests for the calendar endpoint
- Tests cover various scenarios including date filtering, granularity options, and DAG selection
- All existing tests continue to pass
- Static code checks and pre-commit hooks pass

## API Details

**Endpoint:** `GET /ui/calendar/{dag_id}`

**Path Parameters:**
- `dag_id` (string, required): The DAG ID

**Query Parameters:**
- `granularity` (string, optional): Time granularity - `"hourly"` or `"daily"` (default: `"daily"`)
- `logical_date_gte` (datetime, optional): Start date for filtering (ISO 8601 format)
- `logical_date_lte` (datetime, optional): End date for filtering (ISO 8601 format)

**Response Format:**
Returns an array of `CalendarTimeRangeResult` objects with the following structure:

```json
[
  {
    "datetime": "2024-01-15 00:00:00",
    "state": "success",
    "count": 5
  },
  {
    "datetime": "2024-01-15 00:00:00",
    "state": "running",
    "count": 2
  },
  {
    "datetime": "2024-01-16 00:00:00",
    "state": "planned",
    "count": 7
  }
]
```

**Response Fields:**
- `datetime`: The timestamp for the calendar entry
- `state`: Task state - one of `"queued"`, `"running"`, `"success"`, `"failed"`, or `"planned"`
- `count`: Number of task instances in this state at the given time

## Related Links
- [[Discussion](https://github.com/apache/airflow/discussions/50907)](https://github.com/apache/airflow/discussions/50907)
- [[Reference PR](https://github.com/apache/airflow/pull/37988)](https://github.com/apache/airflow/pull/37988) that provided the main logic
- [[Specific changes](https://github.com/apache/airflow/pull/37988/files#diff-917b70d2661fae8322c538a4f60b9d4ba755556036fe0c527eb0411a6ac60f2eL2841)](https://github.com/apache/airflow/pull/37988/files#diff-917b70d2661fae8322c538a4f60b9d4ba755556036fe0c527eb0411a6ac60f2eL2841)

SavingFrame committed:
Add DAG calendar view endpoint

Implement calendar view functionality that provides DAG scheduling
information in a calendar format.
The endpoint allows users to visualize DAG schedules and execution
patterns over time, improving workflow management and scheduling
oversight.
In case if you want get get data for several days you can use daily
granularity.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py:
```diff
@@ -0,0 +1,38 @@
+
+
+class CalendarTimeRangeResult(BaseModel):
+    """Task Instance Summary model for the Grid UI."""
```
Wrong doc string

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/routes/ui/calendar.py:
```diff
@@ -0,0 +1,82 @@
+        granularity: Time granularity ("hourly" or "daily")
+
+    Returns:
+        List of calendar time range results
```
You can remove, this is already documented in swagger from the params

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py:
```diff
@@ -0,0 +1,38 @@
+from airflow.utils.state import DagRunState
+
+
+class CalendarTimeRangeResult(BaseModel):
```
Can you follow the naming convention of other  datamodles please. (Response vs Body)


(Collection, with `total_entries`, even if this endpoint is not paginated)

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py:
```diff
@@ -0,0 +1,38 @@
+class CalendarTimeRangeResult(BaseModel):
+    """Task Instance Summary model for the Grid UI."""
+
+    datetime: datetime
```
nit: I think we should keep 'date' even if its a datetime. (start_date, end_date are all called date but they are actually datetime), just for consistencey

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/services/ui/calendar.py:
```diff
@@ -0,0 +1,321 @@
+        calendar_results = [
+            CalendarTimeRangeResult(
+                # ds.datetime in sqlite and mysql is a string, in postgresql it is a datetime
+                datetime=ds.datetime.replace(tzinfo=None)
```
```suggestion
                datetime=ds.datetime.replace(tzinfo=None)
```

Why setting the tzinfo at None like this? (If there is a tzinfo this will actually be wrong, and the AIP should return tz aware datetime)

SavingFrame replied:
Okay, it was a point i was doubt about. Now I always return datetime in UTC format. My assumption that we always have logical_date in the  UTC in any database, so we can use `date()` function and then convert it into UTC. The do the same with hourly granularity.

Thank you a lot for reviewing. All comments are fixed

pierrejeambrun reviewed (commented):
Nice, thanks for the pull request.

A few suggestions

SavingFrame committed:
restructure calendar response format

Change calendar endpoint to return structured response with total count
and dag runs array instead of flat list for better API consistency.

SavingFrame committed:
Update swagger documentation

SavingFrame committed:
standardize calendar datetime format to ISO 8601 with UTC

Update calendar API to return consistent ISO 8601 datetime format with
UTC timezone suffix across all database dialects and remove timezone
stripping.

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py:
```diff
@@ -0,0 +1,45 @@
+from airflow.utils.state import DagRunState
+
+
+class CalendarTimeRangeDagRuns(BaseModel):
```
```suggestion
class CalendarTimeRangeResponse(BaseModel):
```

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py:
```diff
@@ -0,0 +1,45 @@
+    count: int
+
+
+class CalendarTimeRangeDagRunsResponse(BaseModel):
```
```suggestion
class CalendarTimeRangeCollectionResponse(BaseModel):
```

pierrejeambrun reviewed (approved):
Just a couple of Nit.

Thanks for the PR, I think it's a good start to iterate on. Since it's a UI endpoint we are free to modify this as we see fit.

SavingFrame committed:
Update airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>

SavingFrame committed:
Update airflow-core/src/airflow/api_fastapi/core_api/datamodels/ui/calendar.py

Co-authored-by: Pierre Jeambrun <pierrejbrun@gmail.com>

SavingFrame committed:
Rename pydantic models
